JavaScript (v3): CloudWatch Logs Wkflw - Working snapshot. (#5785)
awsdocs/aws-doc-sdk-examples: workflows/cloudwatch_logs_large_query/README.md

# CloudWatch Logs large query

## Overview

> This example shows how to use AWS SDKs to perform a query on CloudWatch logs and get more than the maximum amount of 10,000 logs back.

Reviewer: ```suggestion
This example shows how to use AWS SDKs to perform a query on CloudWatch logs and get more than the maximum number of 10,000 logs back.
```

> The CloudWatch Logs API is capped at 10,000 records for requests that [read](https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_GetLogEvents.html) or [write](https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_PutLogEvents.html). This example breaks down one query into multiple queries if more than the maximum amount of records is returned from the query.

Reviewer: ```suggestion
The CloudWatch Logs API is capped at 10,000 records for requests that [read](https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_GetLogEvents.html) or [write](https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_PutLogEvents.html). This example breaks down one query into multiple queries if more than the maximum number of records are returned from the query.
```

An AWS CloudFormation [template](./resources/stack.yaml) exists in the [resources folder](./resources/). The template contains two resources: a CloudWatch log group, and a CloudWatch log stream that will be used as the default log group and stream for the queries made in the example. We recommend deploying a stack with this template.

### Prerequisites

If you need to, [install or update the latest version of the AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html).

### Deploy resources

To deploy the stack using the template, run the following command:

```
aws cloudformation deploy --template-file stack.yaml --stack-name LargeQueryStack
```

### Destroy resources

To destroy the stack, run the following command:

```
aws cloudformation delete-stack --stack-name LargeQueryStack
```

## Sample logs

A lot of logs are needed to make a robust example. If you happen to have a log group with over 10,000 logs at the ready, great! If not, there are two resources that can help:

### Prerequisites

If you need to, [install or update the latest version of the AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html).

> 

Reviewer: Is this section needed here? It also appears earlier, and here it gets in between the lead-in to the resources and the list of resources.

Author: Good catch. It's an oversight.

### Resources

> 1. [make-log-files.sh](./resources/make-log-files.sh) will create 50,000 logs and divide them amongst 5 files of 10,000 logs each (the maximum for each call to 'PutLogEvents'). Two timestamps will be output to the console. These timestamps can be used to configure the query. **Five minutes of logs, starting at the time of execution, will be created. Wait at least five minutes after running this script before attempting to query.**

Reviewer: ```suggestion
1. [make-log-files.sh](./resources/make-log-files.sh) will create 50,000 logs and divide them among 5 files of 10,000 logs each (the maximum for each call to 'PutLogEvents'). Two timestamps will output to the console. These timestamps can be used to configure the query. **Five minutes of logs, starting at the time of execution, will be created. Wait at least five minutes after running this script before attempting to query.**
```
